### PR TITLE
Add support for idirafter flag in GCC/Clang

### LINKED ISSUE #2021: Add support for vendored includes on linux.
**What problem will this solve?**
There is a problem with some (most) linux distro-provided mozjs-*/js*package (or whatever spidermonkey (any version) is packaged as on that system), in that it contains a copy of icu headers that don't match the version of icu installed on the system. This mismatch causes a linking issue in [0 A.D.'s Empire Ascendant](play0ad.com) build process, preventing us to use system provided libraries.

**What might be a solution?**
[This revision ](https://code.wildfiregames.com/D3127#change-rhfEiPNVQqRG )contains a possible solution, by telling gcc (and clang, as it follows enough of the same code path within premake) to look through the mozjs-* includes directory after the system standard directories. This means that gcc/clang will look for icu in the system standard location (and find it) before searching the mozjs-* location.

**What other alternatives have you already considered?**
Keeping a bundled copy of the library forever.

### PR DESCRIPTION
**Pull request by nickclark2016**
**What does this PR do?**

Adds support for the `-idirafter` flag with GCC and Clang in the gmake2 and codelite exporters.

**How does this PR change Premake's behavior?**

No changes to the existing Premake behavior.  Introduces `includedirsafter` API.  For GCC and Clang utilizing the gmake2 or Codelite exporters, the directories specified by `includedirsafter` will utilize `-idirafter` flag for includes.  For XCode using GCC or Clang, directories defined by `includedirsafter` will be at the end of the `SYSTEM_HEADER_SEARCH_PATHS` block.  For Visual Studio using MSVC or Clang, directories defined by `includedirsafter` will at the end of the `ExternalIncludePath` element if in VS2019 or later or in the `IncludePath` element if 2017 or before. Similarly, if the MSC toolset is used, `/external:I` will be utilized for `includedirsafter` when using `msc-142` or newer, or `/I` for `msc-141` or earlier.

**Anything else we should know?**

Closes #2021

**Did you check all the boxes?**

- [ ] Focus on a single fix or feature; remove any unrelated formatting or code changes
- [x] Add unit tests showing fix or feature works; all tests pass
- [x] Mention any [related issues](https://github.com/premake/premake-core/issues) (put `closes #XXXX` in comment to auto-close issue when PR is merged)
- [x] Follow our [coding conventions](https://github.com/premake/premake-core/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] Minimize the number of commits
- [x] Align [documentation](https://github.com/premake/premake-core/tree/master/website) to your changes

*You can now [support Premake on our OpenCollective](https://opencollective.com/premake). Your contributions help us spend more time responding to requests like these!*

### CODE REVIEW

File: src/tools/gcc.lua [hunk 1]
```diff
@@ -318,6 +336,11 @@
 			table.insert(result, '-isystem ' .. p.quoted(dir))
 		end
 
+		for _, dir in ipairs(includedirsafter or {}) do
```
Comment (Jarod42) [inline src/tools/gcc.lua:339]:
```suggestion
		for _, dir in ipairs(includedirsafter or cfg.includedirsafter or {}) do
```
"Future" proof for generators not in core (might also apply to previous parameters too).

Comment (nickclark2016) [reply src/tools/gcc.lua:339]: I tend to disagree there. That'll change the behavior of those exporters, potentially in unexpected ways. I see this as a responsibility of the caller, not this function.

Comment (Jarod42) [reply src/tools/gcc.lua:339]: `includedirsafter` would be ignored by unmodified generators.
With my proposal, there is (a big) chance to have expected behavior.
Else it would be broken too (in another way) and requires changes anyway (pass `cfg.includedirsafter` for your version, pass `{}` for mine, and handle `includedirsafter` another way.

codelite/gmake2 might be unmodified ;-)
xcode/visual handle it differently anyway

gmake support seems missing BTW.

Comment (nickclark2016) [reply src/tools/gcc.lua:339]: `includedirsafter` would be ignored by unmodified generators in this implementation, since arguments are optional.  Correct, XCode and Visual Studio handle it differently, as they don't support functionality like `idirafter` provides. GMake support is intentionally missing, as it's deprecated.

If we do change the behavior of how this works, I think it'd be better to have this as a separate PR, as it's out of scope of the current functionality.  As written, the new behavior matches the expected behavior from previous `includedirs` and friends.

Comment (nickclark2016) [reply src/tools/gcc.lua:339]: On top of this, if any 3rd party module passing `nil` to the function with your mechanism to attempt "skip" a parameter (rather than use `{}`) would break, as it would now check `cfg.####` before just putting in `{}`.  Given we're to beta, I'm a bit hesitant to change existing behavior.

Comment (Jarod42) [reply src/tools/gcc.lua:339]: As usual, we disagree ;-) on expected behavior.
But ok.
(I will have to create 3 PRs to add support to the external generators ^_^)

Shouldn't doc specify that it is unsupported by gmake?
(even if deprecated, some features only works with gmake and not gmake2 (D, premake-qt, ...)
(fixing it seems simpler btw).

Comment (nickclark2016) [reply src/tools/gcc.lua:339]: I think the plan is to completely drop gmake support in 5.0 full release.  We can update the documentation to explicitly call out lack of gmake support, though.

Comment (Jarod42) [reply src/tools/gcc.lua:339]: As long as gmake is here, either:
- Update it (add `cfg.includedirsafter` to `toolset.getincludedirs` in gmake)
- Specify in includedirsafter.md than gmake is unsupported.

First solution seems simpler/better to me.

File: tests/tools/test_msc.lua [hunk 2]
```diff
@@ -308,6 +308,23 @@
 		test.contains("/external:I/usr/local/include", msc.getincludedirs(cfg, cfg.includedirs, cfg.externalincludedirs))
 	end
 
+--
+-- Check handling includedirsafter.
+--
+
+function suite.cflags_onIncludeDirsAfter()
+	includedirsafter { "/usr/local/include" }
```
Comment (Jarod42) [inline tests/tools/test_msc.lua:316]: Inhabituel path for msc ^_^
(but ok).

File: website/docs/includedirsafter.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+does not support include directory ordering, these directories are added to the external include directory path.
+
+```lua
+includedirsafter { "paths" }
+```
+
+### Parameters ###
+
+`paths` specifies a list of include file search directories. Paths should be specified relative to the currently running script file.  Search order is evaluated from
+right to left.
```
Comment (Jarod42) [inline website/docs/includedirsafter.md:11]: According to https://gcc.gnu.org/onlinedocs/gcc/Directory-Options.html
it is left-to-right (for gcc at least).

File: website/docs/includedirsafter.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+### Applies To ###
+
+Project configurations.  GCC and Clang are the only toolsets supporting the ordering functionality in the gmake, gmake2 and Codelite exporters.  All exporters and toolsets
```
Comment (Jarod42) [inline website/docs/includedirsafter.md:15]: Shouldn't "remarks" go in Availability?

File: website/sidebars.js [hunk 5]
```diff
@@ -235,6 +236,7 @@ module.exports = {
 						'rules',
 						'runpathdirs',
 						'runtime',
+						'sanitize',
```
Comment (Jarod42) [inline website/sidebars.js:239]: This fix is unrelated to this PR ;-)

File: website/docs/includedirsafter.md (hunk 3 above)
Comment (nickclark2016) [reply website/docs/includedirsafter.md:11]: Good catch. I'm pretty sure it was in my head correct, I just wrote it down wrong.

File: tests/tools/test_msc.lua (hunk 2 above)
Comment (nickclark2016) [reply tests/tools/test_msc.lua:316]: Not gonna lie, just stole these from above it in the file

File: website/sidebars.js (hunk 5 above)
Comment (nickclark2016) [reply website/sidebars.js:239]: Yeah, I just had it locally and didn't catch it.  Want me to back this out and put in a separate PR?

Comment (Jarod42) [reply website/sidebars.js:239]: As you prefer.

but if you keep it, you have to uncheck:
- [ ] Focus on a single fix or feature; remove any unrelated formatting or code changes
 ;-)

Comment (nickclark2016) [reply website/sidebars.js:239]: Unchecked

**Comment by Jarod42** [comment]
For info, UT added in my [repo](https://github.com/Jarod42/premake-sample-projects/tree/includedirsafter_test) (using this branch), and as expected, core-generators pass, external module generators fail (see [premake-sample-projects/actions includedirsafter_test](https://github.com/Jarod42/premake-sample-projects/actions))

**Review by samsinsane — APPROVED** [review]

**Comment by Jarod42** [comment]
@samsinsane : UTs are broken with that merge (due to previous merges).

**Comment by samsinsane** [comment]
@Jarod42 thanks for pointing that out, I'll get that fixed up. Sorry about that!
